File: packages/core/src/provider.ts
import type { CreatePaymentInput, PaymentSession, PaymentStatus, RefundResult, WebhookEvent } from "./types";
import type { Money } from "./money";

export interface PaymentProvider {
  readonly name: string;

  createPayment(input: CreatePaymentInput): Promise<PaymentSession>;

  getPaymentStatus(paymentId: string): Promise<PaymentStatus>;

  /** Not all providers support refunds — check for presence before calling. */
  refund?(paymentId: string, amount?: Money): Promise<RefundResult>;

  /** Not all providers send webhooks — check for presence before calling. */
  verifyWebhook?(payload: unknown, signature: string): Promise<WebhookEvent>;
}

File: sdk-go/provider.go
package openpay

import "context"

// Provider is the interface all payment adapters must implement.
type Provider interface {
	Name() string
	CreatePayment(ctx context.Context, input CreatePaymentInput) (*PaymentSession, error)
	GetPaymentStatus(ctx context.Context, paymentID string) (PaymentStatus, error)
}

// Refunder is implemented by providers that support refunds.
type Refunder interface {
	Refund(ctx context.Context, paymentID string, amount *Money) (*RefundResult, error)
}

// WebhookVerifier is implemented by providers that send signed webhooks.
// payload is the raw request body bytes.
type WebhookVerifier interface {
	VerifyWebhook(ctx context.Context, payload []byte, signature string) (*WebhookEvent, error)
}
